### PR TITLE
OHRM-240:Google OpenID support bug fixed

### PR DESCRIPTION
**Pull request by supunrandunurk**

### CODE REVIEW

**Comment by samanthajayasinghe** [comment]
@supunrandunurk Any reason to close old PR and created a new one? so we lost all review comments.
I strongly disagree to use the same commit message for each commit which makes no clue about commit history. 
FYI : @icechathuranga

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php [hunk 1]
```diff
@@ -23,43 +23,47 @@
...
+    {
         $gClient = new Google_Client();
         $gClient->setApplicationName('Login to Google +');
         $gClient->setClientId($authProvider->getClientId());
         $gClient->setClientSecret($authProvider->getClientSecret());
         $gClient->setRedirectUri($provider->getProviderUrl());
         $gClient->setDeveloperKey($authProvider->getDeveloperKey());
-        $gClient->addScope(array(self::EMAIL_SCOPE,self::PROFILE_SCOPE));
+        $gClient->addScope(array(self::EMAIL_SCOPE, self::PROFILE_SCOPE));
+        $requestCode=$_GET['code'];
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php]: $_GET is not accepted inside service class

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php [hunk 2]
```diff
@@ -23,43 +23,48 @@
  * Description of GoogleAuthProvider
  */
 class GoogleAuthProvider extends AbstractAuthProvider {
-
     const PROFILE_SCOPE = 'profile';
     const EMAIL_SCOPE = 'email';
+    protected $loginService = null;
+    protected $option= null;
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php]: option is an object or string? 
if it's a string then default value should be ''

I feel this should be array and "requestcode" is one parm on it

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php [hunk 3]
```diff
@@ -149,6 +149,8 @@ public function getAuthProviderObj($providerType, $provider, $authProviderDetail
                 break;
             case AuthProviderExtraDetails::GOOGLE_PLUS:
                 $googleAuthProvider = new GoogleAuthProvider();
+                $requestCode=$_GET['code'];
+                $googleAuthProvider->setOption($requestCode);
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php]: may be ['requestcode' => $requestCode ]

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php (hunk 2 above)
Comment (supunrandunurk) [reply symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php]: option is string type variable

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php (hunk 3 above)
Comment (supunrandunurk) [reply symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php]: I change requestCode to Array and assign $_GET values.

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php [hunk 4]
```diff
@@ -23,43 +23,49 @@
...
-        $gClient->addScope(array(self::EMAIL_SCOPE,self::PROFILE_SCOPE));
-
-        $gService = new Google_Service_Plus($gClient);
+        $gClient->addScope(array(self::EMAIL_SCOPE, self::PROFILE_SCOPE));
+        $requestParameters = $this->getOption();
+        $requestCode = $requestParameters['code'];
 
-        if (isset($_GET['code'])) {
-            $gClient->authenticate($_GET['code']);
+        if (isset($requestCode)) {
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php]: requestCode is already initiated .. so why checking again ?

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php [hunk 5]
```diff
@@ -23,43 +23,49 @@
...
             $dataArray['providerid'] = $provider->getProviderId();
-            $dataArray['useridentity'] = $gClient->getAccessToken();
+            $dataArray['useridentity'] = json_encode($gClient->getAccessToken());
             $success = $this->getOpenIdService()->setOpenIdCredentials($username, $dataArray);
             if ($success) {
-                $flag = array('type' => 'true', 'message' => 'User has authentication!');
+                $this->getLoginService()->addLogin();
+                $flag = array(
+                    'type' => 'true',
+                    'message' => 'User has authentication!'
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php]: This is plain English message. What happen for multi languages 
I think it should be __("User has authentication")

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php [hunk 6]
```diff
@@ -23,43 +23,49 @@
...
+                );
                 return $flag;
             } else {
-                $flag = array('type' => 'false', 'message' => 'User Account Not found');
+                $flag = array(
+                    'type' => 'false',
+                    'message' => 'Invalid Credentials : you Have No OpenID account in Orangehrm try loging with OrangeHRM credentials'
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php]: same here

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php [hunk 7]
```diff
@@ -33,39 +34,40 @@ class openIdCredentialsAction extends baseOpenIdAction {
...
-                $provider = $this->getOpenIdProviderService()->getOpenIdProvider($providerId);
-                $authProviderDetails = $this->getAuthProviderExtraDetailsService()->getAuthProviderDetailsByProviderId($providerId);
-                if (($authProviderDetails instanceof AuthProviderExtraDetails)) {
-                    $providerType = $authProviderDetails->getProviderType();
-                    $sfUser->setAttribute('auth.providerType', $providerType);
-                    $flag = $this->getAuthProviderObj($providerType, $provider, $authProviderDetails);
+                    $providerId = $request->getParameter('openIdProvider');
+                    $sfUser->setAttribute('auth.providerId', $providerId);
+                    $provider = $this->getOpenIdProviderService()->getOpenIdProvider($providerId);
+                    $authProviderDetails = $this->getAuthProviderExtraDetailsService()->getAuthProviderDetailsByProviderId($providerId);
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php]: use multi-line for readability

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php [hunk 8]
```diff
@@ -113,22 +120,26 @@ public function getHomePageService() {
...
     }
 
-    public function setHomePageService($homePageService) {
+    public function setHomePageService($homePageService)
+    {
         $this->homePageService = $homePageService;
     }
 
-    public function getSystemUserService() {
+    public function getSystemUserService()
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php]: It's better not to format untouched functions

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php [hunk 9]
```diff
@@ -141,14 +152,16 @@ public function showFlashMessage($flag) {
...
+    public function getAuthProviderObj($providerType, $provider, $authProviderDetails)
+    {
         switch ($providerType) {
             case AuthProviderExtraDetails::OPEN_ID:
                 $openIdAuthProvider = new OpenIdAuthProvider();
                 $flag = $openIdAuthProvider->validateUser($provider);
                 break;
             case AuthProviderExtraDetails::GOOGLE_PLUS:
                 $googleAuthProvider = new GoogleAuthProvider();
+                $googleAuthProvider->setOption($_GET);
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php]: can't we get this as param as $request->getParams() ?

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php (hunk 4 above)
Comment (supunrandunurk) [reply symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php]: Actually, fetchAccessTokenWithAuthCode() function call twice at a time.first time requestCode is empty then isset() return true and call fetchAccessTokenWithAuthCode() function and redirect to same page with URL value. after that requestCode set and ignore the if condition we can use isset() and is_empty()both but google recommend isset();

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php (hunk 9 above)
Comment (supunrandunurk) [reply symfony/plugins/orangehrmOpenidAuthenticationPlugin/modules/openidauth/actions/openIdCredentialsAction.class.php]: baseOpenIdAction  doesn't provide $request->getParamameters() but it provide  $this->getRequest()->getParameter('code'); we can use this.

File: symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php [hunk 10]
```diff
@@ -23,43 +23,51 @@
...
-        $gClient->addScope(array(self::EMAIL_SCOPE,self::PROFILE_SCOPE));
-
-        $gService = new Google_Service_Plus($gClient);
+        $gClient->addScope(array(self::EMAIL_SCOPE, self::PROFILE_SCOPE));
+        $requestParameters = $this->getOption();
+        $requestCode = $requestParameters['code'];
 
-        if (isset($_GET['code'])) {
-            $gClient->authenticate($_GET['code']);
+        if (isset($requestCode)) {
```
Comment (samanthajayasinghe) [inline symfony/plugins/orangehrmOpenidAuthenticationPlugin/lib/authproviders/GoogleAuthProvider.php]: This is not right .. it shild be like this
if($requestParameters['code']) { 
...
}

**Pull request merged**